In [256]:
import pandas as pd 
import numpy as np

In [333]:
fields = ['CALL_NO', 'UNIT_DISPATCH_DATE', 'Time of dispatch', 'CALL_TYPE_FINAL', 'CALL_TYPE_FINAL_D',  'XCOORD', 'YCOORD']

In [371]:
data = pd.read_csv('/home/march/Downloads/EMSDataFixed.csv', usecols=fields)

/home/march/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [372]:
data.isnull().sum(axis=0)

CALL_NO                   0
UNIT_DISPATCH_DATE        0
Time of dispatch      18317
CALL_TYPE_FINAL           0
CALL_TYPE_FINAL_D         0
XCOORD                  619
YCOORD                  619
dtype: int64

In [373]:
# Rename Time of dispatch to DISPATCH_TIME
data.rename(columns={'Time of dispatch':'DISPATCH_TIME'}, inplace=True)

The 18,317 rows (out of 1018169) that are missing time of dispatch are the final 18,317 rows. This corresponds to dates from 11/01/2016 to 12/31/2016. For point process purposes, removing them from use. 

Run this for proof of that fact: 

time_isnull = data['Time of dispatch'].isnull()
lastRead = False
false_true_flip = []
for i in range(0, len(time_isnull)):
    if lastRead != time_isnull[i]:
        false_true_flip.append(i)
    lastRead = time_isnull[i]
print(false_true_flip) # index of flip from False to True = [999852]

In [374]:
# Removing final 18317 rows
indexes_to_remove = []
for i in range(999852, len(data)):
    indexes_to_remove.append(i)
data.drop(data.index[indexes_to_remove], inplace=True)
data.isnull().sum(axis=0)

CALL_NO                 0
UNIT_DISPATCH_DATE      0
DISPATCH_TIME           0
CALL_TYPE_FINAL         0
CALL_TYPE_FINAL_D       0
XCOORD                611
YCOORD                611
dtype: int64

In [375]:
import matplotlib.pyplot as plt
plt.rcParams['figure.figsize'] = [20,10]
loc_null = data['XCOORD'].isnull()
# plt.plot(loc_null, '.')
# The distribution of lost data is *fairly* regular

There are 611 points without any coordinates, which comes out to 100.0*611/999852) = 0.0611% of the rows. As a first pass, deleting these rows from the dataframe as the amount (and distribution) seems small enough not to greatly affect the models. 

In [376]:
data.isna().sum(axis = 0)

CALL_NO                 0
UNIT_DISPATCH_DATE      0
DISPATCH_TIME           0
CALL_TYPE_FINAL         0
CALL_TYPE_FINAL_D       0
XCOORD                611
YCOORD                611
dtype: int64

In [377]:
data.dropna(axis=0, how='any', inplace=True)
data.isna().sum(axis = 0)

CALL_NO               0
UNIT_DISPATCH_DATE    0
DISPATCH_TIME         0
CALL_TYPE_FINAL       0
CALL_TYPE_FINAL_D     0
XCOORD                0
YCOORD                0
dtype: int64

Also need to remove gps points that are obviously wrong. We are removing anything that is smaller than 37 long because that is in Kentucky, not Indiana. 

In [378]:
data = data[data.XCOORD>38]
data.reset_index(drop=True, inplace=True)

In [385]:
# convert dispatch times into strings to manipulate into datetime, next three cells
data.DISPATCH_TIME = data.DISPATCH_TIME.astype(str)

In [453]:
fixed_times = []
for i in range(0, len(data)):
    time = data.DISPATCH_TIME[i]
    if len(time)<8:
        time = time.zfill(6)
        s = time
        t = iter(s)
        time = ':'.join(a+b for a,b in zip(t, t))
    fixed_times.append(time)

In [454]:
data.UNIT_DISPATCH_DATE = data.UNIT_DISPATCH_DATE.astype(str)

In [455]:
from datetime import datetime
datetimes = []
for i in range(0, len(data)):
    datetime_object = datetime.strptime(data.UNIT_DISPATCH_DATE[i] + fixed_times[i], '%Y-%m-%d%H:%M:%S')
    datetimes.append(datetime_object)

In [460]:
datetimes[len(data)-1]

datetime.datetime(2016, 10, 31, 23, 58, 3)

In [461]:
data['Date_Time '] = datetimes

In [462]:
data

CALL_NO UNIT_DISPATCH_DATE DISPATCH_TIME CALL_TYPE_FINAL  \
0        61209515         2006-12-28        044548            773A   
1        61209516         2006-12-28        044853            774A   
2        61209517         2006-12-28        045219            768A   
3        61209519         2006-12-28        051057            773A   
4        61209520         2006-12-28        051756            790A   
5        61209521         2006-12-28        051906            765A   
6        61209522         2006-12-28        053302            773A   
7        61209523         2006-12-28        053916            768A   
8        61209524         2006-12-28        054924            770A   
9        61209525         2006-12-28        060030            790A   
10       61209526         2006-12-28        061515            792B   
11       61209527         2006-12-28        060949            765A   
12       61209528         2006-12-28        061921            765A   
13       61209529         2006-12-28        061946            768A   
14       61209530         2006-12-28        062503            768A   
15       61209532         2006-12-28        062950            770A   
16       61209534         2006-12-28        065645            790B   
17       61209535         2006-12-28        070945            790A   
18       61209536         2006-12-28        071529            792B   
19       61209538         2006-12-28        072300            766A   
20       61209539         2006-12-28        072819            766A   
21       61209542         2006-12-28        075611            766A   
22       61209543         2006-12-28        080339            773A   
23       61209544         2006-12-28        080000            791A   
24       61209545         2006-12-28        080051            775A   
25       61209546         2006-12-28        080117            769B   
26       61209549         2006-12-28        081044            766A   
27       61209550         2006-12-28        081111            790B   
28       61209551         2006-12-28        082139            771A   
29       61209552         2006-12-28        082659            765A   
...           ...                ...           ...             ...   
922509  161015774         2016-10-31      21:57:41            765A   
922510  161015776         2016-10-31      22:03:58            796B   
922511  161015777         2016-10-31      22:07:45            792B   
922512  161015779         2016-10-31      22:10:49            767A   
922513  161015780         2016-10-31      22:16:23           784AU   
922514  161015781         2016-10-31      22:19:34            766A   
922515  161015783         2016-10-31      22:23:42            773A   
922516  161015784         2016-10-31      22:32:17            790B   
922517  161015785         2016-10-31      22:35:36            772A   
922518  161015786         2016-10-31      22:35:42            773A   
922519  161015787         2016-10-31      22:43:02            770A   
922520  161015790         2016-10-31      22:48:33             805   
922521  161015792         2016-10-31      22:55:22           774AU   
922522  161015793         2016-10-31      22:55:31            792A   
922523  161015794         2016-10-31      22:58:35            765A   
922524  161015795         2016-10-31      23:03:41            770A   
922525  161015797         2016-10-31      23:08:22            776A   
922526  161015798         2016-10-31      23:15:03            766A   
922527  161015799         2016-10-31      23:18:03            773A   
922528  161015800         2016-10-31      23:21:20            791A   
922529  161015802         2016-10-31      23:29:13            790A   
922530  161015803         2016-10-31      23:30:41            784C   
922531  161015804         2016-10-31      23:33:08            775A   
922532  161015806         2016-10-31      23:35:01           791AW   
922533  161015807         2016-10-31      23:36:42            768A   
922534  161015809   